### PR TITLE
Reflecting boundary conditions

### PR DESCRIPTION
**Pull request by BenWibking**
Partially fixes https://github.com/parthenon-hpc-lab/athenapk/issues/56.

Only X3-boundaries are implemented. The boundary functions must be manually enrolled by the problem generator.

### CODE REVIEW

File: src/bc.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+#include "bc.hpp"
+
+// Parthenon headers
+#include "main.hpp"
+#include "mesh/mesh.hpp"
+
+using parthenon::Real;
+
+void ReflectingInnerX3(std::shared_ptr<parthenon::MeshBlockData<Real>> &mbd,
```
Comment (pgrete) [inline src/bc.cpp:20]: Any reason to not template this (for different dims) already now?

File: src/bc.hpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  const auto &bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+
+  const auto &range = X1 ? bounds.GetBoundsI(parthenon::IndexDomain::interior)
+                         : (X2 ? bounds.GetBoundsJ(parthenon::IndexDomain::interior)
+                               : bounds.GetBoundsK(parthenon::IndexDomain::interior));
+  const int ref = INNER ? range.s : range.e;
+
+  std::string label = (TYPE == BCType::Reflect ? "Reflect" : "Outflow");
+  label += (INNER ? "Inner" : "Outer");
+  label += "X" + std::to_string(DIR);
```
Comment (pgrete) [inline src/bc.hpp:89]: No need for outflow here but maybe add "AthenaPK" to the label.

File: src/bc.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                        : parthenon::IndexDomain::outer_x3));
+
+  // used for reflections
+  const int offset = 2 * ref + (INNER ? -1 : 1);
+
+  pmb->par_for_bndry(
+      label, nvar, domain, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        if (!q.IsAllocated(l)) return;
+        if (TYPE == BCType::Reflect) {
```
Comment (pgrete) [inline src/bc.hpp:106]: We can probably skip this `TYPE` and just call this  method `ApplyReflectBC``

**Review by pgrete — COMMENTED** [review]
Thanks for porting this over.
In addition to the inline comments I got two more:
a) do we have an "easy" way to test those boundary conditions?
b) some doc on how to enroll those (versus the ones in Parthenon) might be useful.

File: src/bc.cpp (hunk 1 above)
Comment (BenWibking) [reply src/bc.cpp:20]: I can do that.

File: src/bc.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x2)
+                                ? bounds.je(parthenon::IndexDomain::interior)
+                                : bounds.je(parthenon::IndexDomain::entire)},
+      parthenon::IndexRange{IsDomainBound(pmb, parthenon::BoundaryFace::inner_x3)
+                                ? bounds.ks(parthenon::IndexDomain::interior)
+                                : bounds.ks(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x3)
+                                ? bounds.ke(parthenon::IndexDomain::interior)
+                                : bounds.ke(parthenon::IndexDomain::entire)}};
+}
```
[the remark this answers is not in the document]
Comment (BenWibking) [reply src/bc.hpp:53]: This is used in my precipitator pgen. I could move it there if you don't want to include it here.

**Comment by BenWibking** [comment]
For tests, are you thinking of a simulation that would run and then we would check the solution, or just a unit test to check that the ghost cells are filled correctly for a given example of interior cells?
